pdfrw/objects/pdfstring.py

```diff
@@ -1,74 +1,547 @@
+            by changing the bytes_encoding parameter to 'literal' or 'hex'
+            to force a particular conversion method.
+        """
+
+        # If hexadecimal is not being forced, then figure out how long
+        # the escaped literal string will be, and fall back to hex if
+        # it is too long.
+
+        force_hex = bytes_encoding == 'hex'
+        if not force_hex:
+            if bytes_encoding not in ('literal', 'auto'):
+                raise ValueError('Invalid bytes_encoding value: %s'
+                                 % bytes_encoding)
+            splitlist = (cls.escape_splitter or cls.init_escapes())(raw)
+            if bytes_encoding == 'auto' and len(splitlist) // 2 >= len(raw):
+                force_hex = True
+
+        if force_hex:
+            # The spec does not mandate uppercase,
+            # but it seems to be the convention.
+            fmt = '<%s>'
+            result = binascii.hexlify(raw).upper()
+        else:
+            fmt = '(%s)'
+            splitlist[1::2] = [(b'\\' + x) for x in splitlist[1::2]]
+            result = b''.join(splitlist)
+
+        return cls(fmt % convert_load(result))
+
+    @classmethod
+    def from_unicode(cls, source, text_encoding='auto',
+                     bytes_encoding='auto'):
+        """ The from_unicode() constructor is called to encode a source
+            string into a PdfString that is suitable for inclusion in a PDF.
+
+            NOTE:  There is no magic in the encoding process.  A user
+            can certainly do his own encoding, and simply initialize a
+            PdfString() instance with his encoded string.  That may be
+            useful, for example, to add line breaks to make it easier
+            to load PDFs into editors, or to not bother to escape balanced
+            parentheses, or to escape additional characters to make a PDF
+            more readable in a file editor.  Those are features not
+            supported by this method.
+
+            from_unicode() can use a heuristic to figure out the best
+            encoding for the string, or the user can control the process
+            by changing the text_encoding parameter to 'pdfdocencoding'
+            or 'utf16', and/or by changing the bytes_encoding parameter
+            to 'literal' or 'hex' to force particular conversion methods.
+
+            The function will raise an exception if it cannot perform
+            the conversion as requested by the user.
+        """
+
+        # Give preference to pdfdocencoding, since it only
+        # requires one raw byte per character, rather than two.
+        if text_encoding != 'utf16':
+            force_pdfdoc = text_encoding == 'pdfdocencoding'
+            if text_encoding != 'auto' and not force_pdfdoc:
+                raise ValueError('Invalid text_encoding value: %s'
+                                 % text_encoding)
+
+            if not source.startswith(cls.uni_bom):
```

Comment:
Now that I understand what this is doing, I think I see another issue:

```python
PdfString.from_unicode(u'þÿ blah', text_encoding='pdfdocencoding')
```

I'd expect that to raise an exception, but I think it will actually use the UTF-16 encoding.